### PR TITLE
WIP: Added random() methods to several distributions

### PR DESCRIPTION
**Pull request by fonnesbeck**
This is **not ready to merge**, but wanted to share for feedback. I am in the process of adding random number generators for all distributions, starting with the univariate continuous distributions. I've added simple unit tests to most of them -- perhaps not as extensive as we could have, but our tests already run long.

### CODE REVIEW

**Comment by twiecki** [comment]
Would there be any benefit to using Theano `RandomStream`?

**Comment by fonnesbeck** [comment]
I initially thought so, but all we are doing is generating values and not adding nodes to the graph, so figured it would be easier to use Numpy. This is what we do for proposals, and is what @jsalvatier did for the uniform distribution so I just followed suit. If there is a performance (or other) benefit, I am open to reversing course.

I see two main uses for `random()`:
1. Goodness of fit via posterior predictive checks
2. Methods like Approximate Bayesian Computing, where we are just sampling from priors

Neither seem to suggest that we need to use Theano.

**Comment by bjedwards** [comment]
Should `numpy.random.seed` be exposed somewhere in the distributions? It is exposed in `sample.py`, but might be wanted here as well.

File: pymc/distributions/continuous.py [hunk 1]
```diff
@@ -201,6 +212,10 @@ def logp(self, value):
 
         return -log(2 * b) - abs(value - mu) / b
 
+    def random(self, size):
+        u = runiform(-0.5, 0.5, size)
+        return self.mu - np.sign(u)*np.log(1 - 2*np.abs(u))/self.b
```
Comment (jsalvatier) [inline pymc/distributions/continuous.py]: maybe consider just using rexponential and randomizing the sign. Maybe this is faster though.

File: pymc/distributions/continuous.py [hunk 2]
```diff
@@ -286,6 +303,10 @@ def logp(self, value):
             lam > 0,
             nu > 0)
 
+    def random(self, size):
+        return (rnormal(self.mu, 1./np.sqrt(self.tau), size) /
```
Comment (jsalvatier) [inline pymc/distributions/continuous.py]: There's a numpy.random.student_t http://docs.scipy.org/doc/numpy/reference/generated/numpy.random.standard_t.html

File: pymc/tests/test_distributions.py [hunk 3]
```diff
@@ -123,6 +123,12 @@ def test_uniform():
             lambda value, lower, upper: sp.uniform.logpdf(value, lower, upper - lower)
             )
 
+def test_random_uniform():
+    lo, hi = 0, 10
+    with Model() as m:
```
Comment (jsalvatier) [inline pymc/tests/test_distributions.py]: You can create just the distribution like `Uniform.dist(lo, hi).random(size)`

File: pymc/tests/test_distributions.py [hunk 4]
```diff
@@ -123,6 +123,12 @@ def test_uniform():
             lambda value, lower, upper: sp.uniform.logpdf(value, lower, upper - lower)
             )
 
+def test_random_uniform():
```
Comment (jsalvatier) [inline pymc/tests/test_distributions.py]: Thanks for including at least some of the tests to set the standard on how to test these. (slightly worried about sounding passive aggressive here, which is totally not my intent).

I would prefer to somehow compare the KDE or histogram of the values against the pdf but maybe that's too complicated for now.

File: pymc/distributions/continuous.py [hunk 5]
```diff
@@ -149,6 +154,9 @@ def logp(self, value):
             alpha > 0,
             beta > 0)
 
+    def random(self, size):
+        return sbeta.ppf(random(size), self.alpha, self.beta)
```
Comment (jsalvatier) [inline pymc/distributions/continuous.py]: Wait, I actually just thought about this more fully. Does this actually work? self.alpha is often going to be a tensor value, which we probably cannot pass to sbeta.ppf and probably don't want to anyway. I think random has to either be a theano tensor or take parameters.

**Comment by jsalvatier** [comment]
Digging the idea.

File: pymc/tests/test_distributions.py (hunk 4 above)
Comment (fonnesbeck) [inline pymc/tests/test_distributions.py]: Yeah, I'm not close to done. I'm not too worried about which method we use to test; I just want to be able to detect egregious departures from what would be expected. I think checking out the first two moments would be sufficient for most distributions

**Comment by fonnesbeck** [comment]
The above is just a rebase.

File: pymc/distributions/continuous.py (hunk 5 above)
Comment (fonnesbeck) [inline pymc/distributions/continuous.py]: Right, this works when the variable is a prior, but will not when any parents are tensors. I suppose I will have to make this a tensor, then evaluate it to return the sampled values.

Comment (jsalvatier) [inline pymc/distributions/continuous.py]: I think now I want to be more clear on how exactly these functions should be used. What's the purpose and use case?

Comment (fonnesbeck) [inline pymc/distributions/continuous.py]: A couple use cases:
- posterior predictive checks: one way of doing goodness of fit checks is to sample from the posterior predictive distributions of various nodes in a model. 
- approximate Bayesian computing: for some models, likelihood-free methods like ABC are effective for approximating posterior distributions. This involves sampling from the priors of the model and comparing simulated data to the observed data. 

I've been working on this a bit this morning and think I have something that works based on RandomStreams. Stay tuned.

**Comment by fonnesbeck** [comment]
Have a peek at this approach. It now deals with sampling from variables with other variables as parents as well as those with scalar-valued parameters.

I would also like to be able to draw values using a `Point`, so that values for certain variables can be specified, and we aren't just drawing from priors.

**Comment by twiecki** [comment]
>    Goodness of fit via posterior predictive checks

This is a critical feature.

>    Methods like Approximate Bayesian Computing, where we are just sampling from priors

I'm not sure how it'd be useful here as ABC is used when you can't formulate a likelihood?

**Comment by fonnesbeck** [comment]
ABC can be used whenever the likelihood is expensive to calculate, not necessarily when you cant formulate it. So it can be useful with large data sets.

**Comment by twiecki** [comment]
Right, but I don't think that's the case for any of the likelihoods we have. In any case though, the posterior predictive stuff is definitely worthwhile.

**Comment by fonnesbeck** [comment]
All of our likelihoods are expensive to calculate if you have enough data. It's not related to any particular implementation, it's just a shortcoming of mcmc in general. It does not scale well with the size of the data set.

**Comment by twiecki** [comment]
Good point. Although one could just compute the summary statistics for exponential families analytically for "non-simulation ABC".

**Comment by fonnesbeck** [comment]
The current implementation (in `continuous.py` only right now) has `random` methods with an optional `point` argument. If left blank, a sample will be drawn from the current variable, conditional on prior draws from its parents (or on the value if it is a prior distribution) -- essentially a draw from the prior predictive distribution.  If a Point is passed, it will draw values based on any parents' values specified in the point. This can allow for any nodes above the parameter in the graph to be fixed at arbitrary values prior to drawing a random value. Happy to turn `draw_values` into a proper function if very long lambda functions offend anyone.

**Comment by twiecki** [comment]
Since this is becoming more relevant, what are the missing steps here?

**Comment by twiecki** [comment]
@fonnesbeck I rebased this branch off of master and force pushed (so make sure to force pull next time).

**Comment by twiecki** [comment]
Current tests fail with:

```
======================================================================
ERROR: pymc3.tests.test_diagnostics.test_gelman_rubin
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/pymc-devs/pymc/pymc3/tests/test_diagnostics.py", line 13, in test_gelman_rubin
    random_seed=[1, 3])
  File "/home/travis/build/pymc-devs/pymc/pymc3/sampling.py", line 88, in sample
    return sample_func(*sample_args)
  File "/home/travis/build/pymc-devs/pymc/pymc3/sampling.py", line 209, in _mp_sample
    traces = p.map(argsample, args)
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/multiprocessing/pool.py", line 251, in map
    return self.map_async(func, iterable, chunksize).get()
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/multiprocessing/pool.py", line 558, in get
    raise self._value
PicklingError: Can't pickle <type 'instancemethod'>: attribute lookup __builtin__.instancemethod failed
```

**Comment by twiecki** [comment]
These are the failing tests, seem to be because of `TransformedVar`:

``` python
======================================================================
ERROR: pymc3.tests.test_distributions.test_random_uniform
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/pymc-devs/pymc3/pymc3/tests/test_distributions.py", line 144, in test_random_uniform
    r = u.random()
AttributeError: 'TransformedRV' object has no attribute 'random'
======================================================================
ERROR: pymc3.tests.test_distributions.test_random_normal
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/pymc-devs/pymc3/pymc3/tests/test_distributions.py", line 190, in test_random_normal
    r = n.random()
TypeError: 'NoneType' object is not callable
======================================================================
ERROR: pymc3.tests.test_distributions.test_random_beta
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/pymc-devs/pymc3/pymc3/tests/test_distributions.py", line 207, in test_random_beta
    r = b.random()
AttributeError: 'TransformedRV' object has no attribute 'random'
======================================================================
ERROR: pymc3.tests.test_distributions.test_random_exponential
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/pymc-devs/pymc3/pymc3/tests/test_distributions.py", line 222, in test_random_exponential
    r = x.random()
AttributeError: 'TransformedRV' object has no attribute 'random'
======================================================================
ERROR: pymc3.tests.test_distributions.test_random_lognormal
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/pymc-devs/pymc3/pymc3/tests/test_distributions.py", line 257, in test_random_lognormal
    r = ln.random(size=10000)
AttributeError: 'TransformedRV' object has no attribute 'random'
```

**Comment by fonnesbeck** [comment]
I'll need to look at this again and re-evaluate. Changed title to WIP.

**Comment by mwibrow** [comment]
@fonnesbeck I'd had a go at some `random` methods (also `pdf` and `cdf` methods) for most of the continuous distributions before I noticed this thread. 

My stuff is way too scrappy for a pull request (no formal tests yet, I've just been doing some ad-hoc visual checks) but it is over at: https://github.com/mwibrow/pymc3/tree/more_rvs/pymc3

Some of it is just wrappers around numpy/scipy stuff, but I've incorporated a variant of the `draw_values` function. 

I guess work is now being duplicated in this area, so not sure how to proceed.

**Comment by twiecki** [comment]
@mwibrow I had a look at your branch and I think you are being too modest. This is a significant piece of work that would certainly make a great contribution. I think unittests shouldn't be too hard to come by and mostly compare to the numpy/scipy implementations as ground truth.

**Comment by twiecki** [comment]
Moreover, it seems like your changes are much closer and more advanced to what we have here currently. Would you be interested in doing a PR that supersedes this one?

**Comment by hadim** [comment]
Any news here ?

**Comment by fonnesbeck** [comment]
Closing this in favor of @mwibrow 's PR in #784

**Pull request closed without merging**